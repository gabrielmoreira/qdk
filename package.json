{
  "name": "qdk",
  "version": "0.3.2",
  "description": "QDK (Quick Development Kit) is a tooling solution for generating configuration files via TypeScript, inspired by Projen.",
  "homepage": "https://github.com/gabrielmoreira/qdk#readme",
  "bugs": {
    "url": "https://github.com/gabrielmoreira/qdk/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/gabrielmoreira/qdk.git"
  },
  "license": "MIT",
  "author": "Gabriel Moreira <gabrielmoreira@gmail.com>",
  "type": "module",
  "bin": "./dist/lib/cli.js",
  "main": "./dist/cjs/index.js",
  "types": "./dist/lib/index.d.ts",
  "exports": {
    ".": {
      "import": "./dist/lib/index.js",
      "require": "./dist/cjs/index.js",
      "types": "./dist/lib/index.d.ts"
    }
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "prebuild": "nr clean",
    "build": "nr typescript:prod:build",
    "clean": "rimraf .nyc_output coverage dist",
    "eslint": "eslint",
    "eslint:fix": "eslint --fix",
    "format": "prettier . --write",
    "prepack": "nr build",
    "prepare": "simple-git-hooks",
    "publish": "clean-publish",
    "next-version": "bumpp --no-commit --no-tag --no-push",
    "prerelease": "bumpp",
    "release": "nr publish",
    "sample:monorepo": "tsx test/samples/monorepo.qdk.ts",
    "test": "vitest",
    "typescript:prod:build": "tsc -p tsconfig.production.json && tsc -p tsconfig.production.json --moduleResolution node --module commonjs --outdir ./dist/cjs",
    "typescript:dev:build": "tsc -p tsconfig.json",
    "typescript:check": "tsc -p tsconfig.json --noEmit --explainFiles > explainTypes.txt"
  },
  "simple-git-hooks": {
    "pre-commit": "npx nr eslint"
  },
  "dependencies": {
    "@types/debug": "^4.1.12",
    "@types/lodash-es": "^4.17.12",
    "@types/vinyl": "^2.0.12",
    "clipanion": "^4.0.0-rc.4",
    "debug": "^4.3.7",
    "find-up-simple": "^1.0.0",
    "hookable": "^5.5.3",
    "immer": "^10.1.1",
    "lodash-es": "^4.17.21",
    "magicast": "^0.3.5",
    "normalize-package-data": "^7.0.0",
    "nr": "^0.1.0",
    "prettier": "^3.3.3",
    "sort-package-json": "^2.10.1",
    "tree-console": "^1.2.1",
    "typanion": "^3.14.0",
    "type-fest": "^4.26.1",
    "vinyl": "^3.0.0",
    "yaml": "^2.5.1"
  },
  "devDependencies": {
    "@antfu/ni": "^0.23.0",
    "@eslint/js": "^9.11.1",
    "@rollup/plugin-typescript": "^12.1.0",
    "@tsconfig/node20": "^20.1.4",
    "@types/node": "^20.16.10",
    "@types/normalize-package-data": "^2.4.4",
    "bumpp": "^9.6.1",
    "clean-publish": "^5.0.0",
    "eslint": "^9.11.1",
    "eslint-config-prettier": "^9.1.0",
    "eslint-plugin-prettier": "^5.2.1",
    "globals": "^15.10.0",
    "memfs": "^4.12.0",
    "prettier-plugin-organize-imports": "^4.1.0",
    "rimraf": "^6.0.1",
    "simple-git-hooks": "^2.11.1",
    "tsx": "^4.19.1",
    "typescript": "^5.6.2",
    "typescript-eslint": "^8.8.0",
    "vitest": "^2.1.2"
  }
}
