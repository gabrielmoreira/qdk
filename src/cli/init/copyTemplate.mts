import { relativeToCwd } from '#qdk';
import Debug from 'debug';
import { globby } from 'globby';
import fs, { existsSync } from 'node:fs';
import { mkdir, readFile, writeFile } from 'node:fs/promises';
import { dirname, join, relative } from 'node:path';
const debug = Debug('qdk:template');

export async function copyTemplate(
  template: string,
  { cwd: destdir, forceOverwrite }: { cwd: string; forceOverwrite: boolean },
) {
  const srcdir = join(import.meta.dirname, '..', '..', 'templates', template);
  debug('srcdir', srcdir);
  const paths = await globby(`**/*`, { cwd: srcdir, dot: true, fs });
  debug('paths', srcdir);
  const directories = [
    ...new Set(paths.map(path => join(destdir, dirname(path)))),
  ];
  debug('directories', directories);
  await Promise.all(directories.map(dir => mkdir(dir, { recursive: true })));
  const copyList = paths.map(path => [join(srcdir, path), join(destdir, path)]);
  debug('copy list', copyList);
  const generated: string[] = [];
  await Promise.all(
    copyList.map(async ([srcpath, destpath]) => {
      debug('Copy from', srcpath, 'to', destpath);
      if (!existsSync(destpath) || forceOverwrite) {
        await copyFile(srcpath, destpath);
        generated.push(relative(destdir, destpath));
      } else {
        console.warn(
          'File',
          relative(destdir, destpath),
          'already exists. Use -f to overwrite it.',
        );
      }
    }),
  );
  return generated;
}

async function copyFile(from: string, to: string) {
  const input = await readFile(from);
  const output = normalizeContent(from, input);
  await mkdir(dirname(to), { recursive: true });
  await writeFile(to, output);
}

const normalizePaths = /(\.js|\.ts|\.cjs|\.mjs|\.mts|\.cts|\.jsx|\.tsx)$/;

function normalizeContent(path: string, content: Buffer) {
  if (!normalizePaths.test(path)) {
    debug('Path ', relativeToCwd(path), 'should not be normalized. Skipping.');
    return content;
  }
  debug('Path ', relativeToCwd(path), 'should be normalized. Processing...');
  return content
    .toString()
    .replaceAll(`'#qdk'`, `'qdk'`)
    .replaceAll(`"#qdk"`, `"qdk"`);
}
