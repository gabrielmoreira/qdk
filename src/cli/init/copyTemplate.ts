import { globby } from 'globby';
import { existsSync } from 'node:fs';
import { copyFile, mkdir } from 'node:fs/promises';
import { dirname, join, relative } from 'node:path';
import { metaDirname } from './dirname.cjs';

export async function copyTemplate(
  template: string,
  { cwd: destdir, forceOverwrite }: { cwd: string; forceOverwrite: boolean },
) {
  const srcdir = join(metaDirname, 'templates', template);
  //console.log('srcdir', srcdir);
  const paths = await globby(`**/*`, { cwd: srcdir, dot: true });
  //console.log('paths', srcdir);
  const directories = [
    ...new Set(paths.map(path => join(destdir, dirname(path)))),
  ];
  //console.log('directories', directories);
  await Promise.all(directories.map(dir => mkdir(dir, { recursive: true })));
  const copyList = paths.map(path => [join(srcdir, path), join(destdir, path)]);
  //console.log('copy list', copyList);
  const generated: string[] = [];
  await Promise.all(
    copyList.map(async ([srcpath, destpath]) => {
      /*console.log(
        'Copy',
        srcpath,
        'to',
        destpath,
        'exists:',
        existsSync(destpath),
        'forceOverwrite',
        forceOverwrite,
      );
      */
      if (!existsSync(destpath) || forceOverwrite) {
        await copyFile(srcpath, destpath);
        generated.push(relative(destdir, destpath));
      } else {
        console.warn(
          'File',
          relative(destdir, destpath),
          'already exists. Use -f to overwrite it.',
        );
      }
    }),
  );
  return generated;
}
