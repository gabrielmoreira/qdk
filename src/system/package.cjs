import { dirname, join } from 'node:path';

export const qdkPackageRoot = join(
  // eslint-disable-next-line no-undef
  dirname(require.resolve('qdk/package.json')),
);
