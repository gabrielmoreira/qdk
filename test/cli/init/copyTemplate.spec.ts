import { vol } from 'memfs';
import { join } from 'node:path';
import { beforeEach, describe, expect, it, vi, vitest } from 'vitest';
import { copyTemplate } from '../../../src/cli/init/copyTemplate.js';
import { qdkPackageRoot } from '../../../src/system/package.cjs';
import { reset, resetFilesystem } from '../../test-helpers.js';

vitest.mock('fs', async () => {
  return await vi.importActual('memfs');
});

vitest.mock('');

vitest.mock('fs/promises', async () => {
  return (await vi.importActual('memfs')).promises;
});
vitest.mock('@/system/package.cjs', () => {
  return {
    qdkPackageRoot: '/pkg/qdk',
  };
});

describe('copyTemplate', () => {
  beforeEach(() => reset());

  it('can copy a blank template', async () => {
    // Given
    resetFilesystem({
      json: {
        // create a fake template
        [join(qdkPackageRoot, '/templates/blank/qdk.config.js')]: 'something',
        [join(qdkPackageRoot, '/templates/blank/.qdk/components/config.ts')]:
          'config',
      },
    });
    // When
    await copyTemplate('blank', { cwd: '/tmp', forceOverwrite: false });
    // Then
    expect(vol.toJSON()).toMatchSnapshot();
  });
});
