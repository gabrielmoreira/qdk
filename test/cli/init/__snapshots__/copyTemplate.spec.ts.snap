// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`copyTemplate > can copy a blank template 1`] = `
{
  "/pkg/qdk/templates/blank/.qdk/components/config.ts": "config",
  "/pkg/qdk/templates/blank/qdk.config.js": "something",
  "/tmp/.qdk/components/config.ts": "config",
  "/tmp/qdk.config.js": "something",
}
`;
